derp/derp_server.go

```diff
@@ -795,6 +825,54 @@ func (c *sclient) run(ctx context.Context) error {
 		if err != nil {
 			return err
 		}
+		c.cleanFlows()
+	}
+}
+
+// flowsTimerFunc tells all the sclients to scan for inactive flows
+// and free them.
+func (s *Server) flowsTimerFunc() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.closed {
+		return
+	}
+	for c := range s.flows {
+		c.flowsCleanNow = true
+	}
+	s.flowsTimer = time.AfterFunc(flowsCleanInterval, s.flowsTimerFunc)
+}
+
+// cleanFlows scans all the flows for this sclient and frees any that
+// have been inactive for more than flowsCleanInterval.
+func (c *sclient) cleanFlows() {
+	if !c.flowsCleanNow {
+		return
+	}
+	c.flowsCleanNow = false
+	c.debugLogf("cleaning flows")
+	now := time.Now()
+	s := c.s
+	for dstKey, meshMap := range c.flows {
+		for srcOrFwd, flow := range meshMap {
+			if now.Sub(flow.lastActive) > flowsInactiveTime {
+				c.debugLogf("deleting flow %s", dstKey.ShortString())
+				// Common case is only one flow for
+				// dstKey, in which case we should
+				// delete the whole submap.
+				if len(c.flows[dstKey]) == 1 {
+					delete(c.flows, dstKey)
+					s.mu.Lock()
+					delete(s.flows[c], dstKey)
+					s.mu.Unlock()
+				} else {
+					delete(c.flows[dstKey], srcOrFwd)
+					s.mu.Lock()
+					delete(s.flows[c][dstKey], srcOrFwd)
+					s.mu.Unlock()
+				}
```

Comment:
It's really hard for me to convince myself that this locking is correct. It looks kinda bizarre.